User registration#8

asmolentzov opened the pull request:
### User Story 8: User Registration
Closes #8 

- Adds a nav bar at the top of the page
- Adds a sign up link
- Adds a sign up page and form
- Adds event listener to form submission to: 
  - Do some basic password validations (make sure the fields are filled out and the passwords match)
  - Send a POST request to the API to create a user
  - Saves the returned API key into the document cookie
  - Redirect the user to the main page
  - Display a success message

#### Review Request
@kaohman Thank you for looking this over!
This PR also pulled in the branch from my other outstanding PR, so you can disregard the Background Image changes. 
I have a few specific questions, but would welcome any feedback!
1. I already have a couple of functions in my `index.js` relating to users, and I suspect I'll have more as I implement Login and Favorite-ing locations. Would you recommend creating a User object or similar to manage that functionality? Is that pretty standard? 
2. I implemented super-basic password checking (basically just making sure they are the same and not blank - the actual authentication takes place on the BE) - are there any shortcuts for this kind of thing, or resources you'd recommend checking out? 
3. I needed a way to persist the API key that the BE returns so that it can be used to keep a user 'logged in'. Is using `document.cookie` a reasonable way to do this? I figure that way I can access it in future requests when I need to send it along, e.g. to 'favorite' a location. 

#### Additional Context
[Project Spec](http://backend.turing.io/module4/projects/self_directed_fe/self_directed_fe_spec)
[Deployed API](https://fierce-meadow-77109.herokuapp.com/api/v1/gifs?location=denver,co)

#### Screenshot
![image](https://user-images.githubusercontent.com/6686861/54329940-7f736700-45d9-11e9-8105-04af5b738ea6.png)

asmolentzov committed:
Adds sign up form page

asmolentzov committed:
Adds CSS ID to location submission form so JS doesn't fire for any form

asmolentzov committed:
Adds basic password validation

asmolentzov committed:
Adds post to users endpoint

asmolentzov committed:
Adds returned API key to document cookie

asmolentzov committed:
Refactors functionality out of event listener for creating user

asmolentzov committed:
Adds alerts to both pages, redirects to main page after successful login

asmolentzov committed:
Runs npm build

asmolentzov committed:
Adds nav bar styling

asmolentzov committed:
Adds styling to sign up form

asmolentzov committed:
Runs npm build

kaohman commented on lib/index.js:
```diff
@@ -1,19 +1,48 @@
+  $('#sign-up').on('submit', registerUser);
+})
+
+const registerUser = (event) => {
```
Related to item 1 in your PR, I'm not sure if you have worked with [classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) at all but creating a User class in a separate file would work well here and you can move your user related functions to be methods on that class like you said. All you need to do in your index.js file then is create a new instance of the class User: `const user = new User(email, password, confPassword)`. Happy to talk through how to structure that class more if you'd like.

kaohman commented on lib/index.js:
```diff
@@ -1,19 +1,48 @@
+
+const createUser = (email, password, confPassword) => {
+  const usersUrl = API_URL + "/users"
+  $.post(usersUrl, { email: email, password: password, password_confirmation: confPassword })
```
If you have a key that matches the name of the existing variable in your code, you can simplify how you create the object to `{ email, password, password_confirmation: confPassword }`. Newer JS syntax allows this, just a nice way to reduce how much typing you have to do and it is more readable.

asmolentzov replied:
NICE!! That's awesome.

kaohman commented on lib/index.js:
```diff
@@ -1,19 +1,48 @@
+  const email = $('input')[0].value;
+  const password = $('input')[1].value;
+  const confPassword = $('input')[2].value;
+  if(!password || !confPassword || password !== confPassword) {
```
Your password verification here makes sense to me and is totally valid as is. A couple other thoughts if you want to make it even more user-friendly...

One thing you could do to simplify this statement is on your `<input>` elements in your HTML, you can add a field of `required` which doesn't allow the user to submit the form if those parameters are empty. Then you don't need to worry about checking if they are blank in your Javascript.

For checking if the password and confPassword fields match, you could also add on change event listeners to your password and confPassword elements to do "live" check if they match as the user types and don't allow the user to submit the form until they match.

asmolentzov replied:
Ah, the `required` makes it so much cleaner! Thank you!

kaohman commented on main.bundle.js:
```diff
@@ -52,19 +52,48 @@
+	var createUser = function createUser(email, password, confPassword) {
+	  var usersUrl = ("http://localhost:3000/api/v1") + "/users";
+	  $.post(usersUrl, { email: email, password: password, password_confirmation: confPassword }).done(function (data) {
+	    document.cookie = 'api_key=' + data.api_key;
```
I've actually never used document.cookie before! In looking at the docs, this looks like the perfect use case. 

We have primarily been using [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) to save this type of information... but we also haven't been working with logins very much since we are just learning how to build a backend now. 

I wouldn't recommend using localStorage though, your method seems more secure. I'm definitely going to look into adding it on my BikeSee project since I have a whole extra week to work in it now 😃!

asmolentzov replied:
🚲 🍪 😃

kaohman commented on main.bundle.js:
```diff
@@ -52,19 +52,48 @@
+	  if (!password || !confPassword || password !== confPassword) {
+	    displayError("Error: Please check passwords");
+	  } else {
+	    createUser(email, password, confPassword);
```
This is where you would add your new instance of the class User if you decide to go the route of separating it into a class.

kaohman commented on lib/index.js:
```diff
@@ -25,4 +54,22 @@ const setCurrentWeather = (weatherInfo) => {
   $('#current-icon').text('').append(currentWeather.icon);
+};
+
+const setBackgroundImage = (imageInfo) => {
```
I would suggest not changing the css directly in your JavaScript file to better follow single responsibility principle. This way all of your styling is all contained in your scss files and is easier to read and debug. Instead, I would look into using [.addClass()](https://api.jquery.com/addclass/) and [.removeClass()](https://api.jquery.com/removeclass/) to your HTML elements and then putting your styling for those classes in your scss files.

asmolentzov replied:
This is a really good point, thank you! I've got another PR open right now and will make this change in there.

asmolentzov replied:
Question about this actually - since the URL of the image changes dynamically depending on what image my API request returns, do you know if I am able to use `addClass` but also somehow pass it the attribute of the url it should be using in this particular instance? 
-- Never mind, I figured out a way to work around this!

kaohman replied:
Ahh okay, your original way might be the best then, if the workaround was difficult. Background images are tricky. I'm so used to React now where you are set your HTML directly in your JS files so it's more straightforward to create an `<img>` tag and set the URL src. That could be another option, to append a new `<img>` HTML element to the DOM and style it as `position: relative` to overlay other elements on top of it.

kaohman commented on lib/index.js:
```diff
@@ -25,4 +54,22 @@ const setCurrentWeather = (weatherInfo) => {
+  console.log(error);
+};
+
+const errorBackground = (error) => {
```
See comment previous comment.

kaohman commented on sign-up.html:
```diff
@@ -0,0 +1,46 @@
+    <div class="container">
+      <form id="sign-up">
+        <h2>Create a New Account</h2>
+        <ul>
```
I think using an unordered list in your form here isn't necessary and might make styling more of a pain than it needs to be. The labels and inputs should sit nicely the way you want them without the list.

asmolentzov replied:
You're totally right! Yay simplifying things!

kaohman reviewed (commented):
Please see my inline comments for suggestions on specific items. I enjoyed reading code from someone who has a different perspective coming from the backend side of things! I really like your cookie implementation to keep the user logged in.

asmolentzov committed:
Removes list formatting from sign-in form and unnecessary styling

asmolentzov committed:
Simplifies createUser post object syntax

asmolentzov committed:
Adds required fields in sign-up form to simplify checking logic

asmolentzov committed:
Runs npm build

asmolentzov commented:
@kaohman Thank you so much for your SUPER helpful comments!! I went ahead and implemented a few changes: 
1. Removing the list formatting from the sign-in form (you're right, it wasn't necessary!) and some then-extraneous styling.
2. Simplifies the post object syntax in the createUser function by leveraging that you don't need to repeat the key/value names if they're the same (e.g. `email: email`). 
3. Adds `required` to the sign-up form fields to remove that logic-checking from the JS.

I also added a few things to my future-follow-up list: 
1. Implement a User class to abstract some of that behavior
2. Add event listeners to my sign up form to implement more 'live' checking for the passwords matching

Let me know if there's anything else you'd suggest, thank you!

asmolentzov committed:
Merge branch 'master' of github.com:asmolentzov/sweater-weather-fe into user-registration#8

asmolentzov committed:
Adds background gradient to sign up page

asmolentzov committed:
Runs npm build
